crates/client-sdk/src/transaction_builder.rs

```diff
@@ -0,0 +1,257 @@
+            .push(action.as_blob(contract_name, caller, callees));
+        Ok(self.runners.last_mut().unwrap())
+    }
+
+    pub fn build<S: StateUpdater>(&mut self, full_states: &mut S) -> Result<BuildResult> {
+        let mut outputs = vec![];
+        for runner in self.runners.iter_mut() {
+            let on_chain_state = self
+                .on_chain_states
+                .get(&runner.contract_name)
+                .cloned()
+                .ok_or(anyhow::anyhow!("State not found"))?;
+            let full_state = full_states.get(&runner.contract_name)?.clone();
+
+            runner.contract_input.blobs = self.blobs.clone();
+            runner.contract_input.private_blob = runner.private_blob(full_state.clone())?;
+            runner.contract_input.initial_state = on_chain_state.clone();
+
+            let out = runner.execute()?;
+            self.on_chain_states
+                .entry(runner.contract_name.clone())
+                .and_modify(|v| *v = out.next_state.clone());
+
+            if let Some(off_chain_new_state) = runner.callback(full_state)? {
+                full_states.update(&runner.contract_name, off_chain_new_state)?;
+            } else {
+                full_states.update(&runner.contract_name, out.next_state.clone())?;
+            }
+
+            outputs.push((runner.contract_name.clone(), out));
+        }
+
+        Ok(BuildResult {
+            identity: self.identity.clone(),
+            blobs: self.blobs.clone(),
+            outputs,
+        })
+    }
+
+    /// Returns an iterator over the proofs of the transactions
+    /// In order to send proofs when they are ready, without waiting for all of them to be ready
+    /// Example usage:
+    /// for (proof, contract_name) in transaction.iter_prove() {
+    ///    let proof: ProofData = proof.await.unwrap();
+    ///    ctx.client()
+    ///        .send_tx_proof(&hyle::model::ProofTransaction {
+    ///            blob_tx_hash: blob_tx_hash.clone(),
+    ///            proof,
+    ///            contract_name,
+    ///        })
+    ///        .await
+    ///        .unwrap();
+    ///}
+    pub fn iter_prove<'a>(
+        &'a self,
+    ) -> impl Iterator<
+        Item = (
+            Pin<Box<dyn std::future::Future<Output = Result<ProofData>> + Send + 'a>>,
+            ContractName,
+        ),
+    > + 'a {
+        self.runners.iter().map(|runner| {
+            let future = runner.prove();
+            (
+                Box::pin(future)
+                    as Pin<Box<dyn std::future::Future<Output = Result<ProofData>> + Send + 'a>>,
+                runner.contract_name.clone(),
+            )
+        })
+    }
+}
+
+pub struct ContractRunner {
+    pub contract_name: ContractName,
+    binary: &'static [u8],
+    contract_input: ContractInput,
+    offchain_cb: Option<Box<dyn Fn(StateDigest) -> Result<StateDigest> + Send + Sync>>,
+    private_blob_cb: Option<Box<dyn Fn(StateDigest) -> Result<BlobData> + Send + Sync>>,
+}
```

Comment:
Second suggestion is to split this off in two: a `ContractMetadata` with name/binary/callbacks, and a `ContractRunner` struct in charge of storing the created `contract_input` and the matching `HyleOutput` and actually executing stuff.

This way if we end up with several blobs for the same contract we don't duplicate the callbacks / makes more sense to me overall in terms of what the structs do (especially with the change above).
IMO.